iocuddle/src/lib.rs

```diff
@@ -0,0 +1,372 @@
+    ///
+    /// Further, you are responsible to ensure that the argument `type` itself
+    /// provides appropriate safe wrappers around its raw contents. For some
+    /// `type`s none are required. For others, particularly `type`s that pass
+    /// pointers to the kernel as `u64`, you need to ensure that things like
+    /// lifetimes are correct.
+    pub const unsafe fn classic(request: c_ulong) -> Self {
+        Self(request, PhantomData)
+    }
+}
+
+impl Ioctl<Read, c_void> {
+    /// Issue an `ioctl` to read a file descriptor's metadata as `c_uint`.
+    ///
+    /// No argument is supplied to the internal `ioctl()` call. The raw
+    /// (positive) return value from the internal `ioctl()` call is returned
+    /// on success.
+    pub fn ioctl(self, fd: &impl AsRawFd) -> Result<c_uint> {
+        let r = unsafe { ioctl(fd.as_raw_fd(), self.0, null::<c_void>()) };
+
+        r.try_into().or_else(|_| Err(Error::last_os_error()))
+    }
+}
+
+impl<T> Ioctl<Read, &T> {
+    /// Issue an `ioctl` to read a file descriptor's metadata as type `T`.
+    ///
+    /// A zeroed instance of type `T` is passed as the first argument to the
+    /// internal `ioctl()` call. Upon success, returns the raw (positive)
+    /// return value and the instance of `T`.
+    pub fn ioctl(self, fd: &impl AsRawFd) -> Result<(c_uint, T)> {
+        let mut out: T = unsafe { core::mem::MaybeUninit::zeroed().assume_init() };
+
+        let r = unsafe { ioctl(fd.as_raw_fd(), self.0, &mut out as *mut _, null::<c_void>()) };
+
+        r.try_into()
+            .and_then(|x| Ok((x, out)))
+            .or_else(|_| Err(Error::last_os_error()))
+    }
+}
+
+impl Ioctl<Write, c_void> {
+    /// Issue an `ioctl` to modify a file descriptor
+    ///
+    /// No argument is provided.
+    ///
+    /// On success, returns the (positive) return value.
+    pub fn ioctl(self, fd: &mut impl AsRawFd) -> Result<c_uint> {
+        let r = unsafe { ioctl(fd.as_raw_fd(), self.0, null::<c_void>()) };
+
+        r.try_into().or_else(|_| Err(Error::last_os_error()))
+    }
+}
+
+impl Ioctl<Write, c_int> {
+    /// Issue an `ioctl` to modify a file descriptor
+    ///
+    /// A C-integer argument is provided.
+    ///
+    /// On success, returns the (positive) return value.
+    pub fn ioctl(self, fd: &mut impl AsRawFd, data: c_int) -> Result<c_uint> {
+        let r = unsafe { ioctl(fd.as_raw_fd(), self.0, data, null::<c_void>()) };
+
+        r.try_into().or_else(|_| Err(Error::last_os_error()))
+    }
+}
+
+impl<T> Ioctl<Write, &T> {
+    /// Issue an `ioctl` to modify a file descriptor
+    ///
+    /// A reference to an immutable instance of `T` is provided as the argument.
+    ///
+    /// On success, returns the (positive) return value.
+    pub fn ioctl(self, fd: &mut impl AsRawFd, data: &T) -> Result<c_uint> {
+        let r = unsafe { ioctl(fd.as_raw_fd(), self.0, data as *const _, null::<c_void>()) };
+
+        r.try_into().or_else(|_| Err(Error::last_os_error()))
```

Comment:
Could be simplified to `.map_err(|_| Error::last_os_error())`

Reply:
Fixed.